dali/kernels/imgproc/roi_test.cc

```diff
@@ -0,0 +1,155 @@
+    TensorShape<3> ref_sh = {3, 3, 3};
+    ASSERT_EQ(ref_sh, sh);
+  }
+  {
+    Roi roi{{0, 2},
+            {5, 6}};
+    auto sh = ::dali::kernels::ShapeFromRoi(roi, 666);
+    TensorShape<3> ref_sh = {4, 5, 666};
+    ASSERT_EQ(ref_sh, sh);
+  }
+  {
+    Roi roi{{0, 0},
+            {0, 0}};
+    auto sh = ::dali::kernels::ShapeFromRoi(roi, 666);
+    TensorShape<3> ref_sh = {0, 0, 666};
+    ASSERT_EQ(ref_sh, sh);
+  }
+}
+
+
+TEST(RoiTest, roi_to_TensorListShape) {
+  using Rois = std::vector<Roi>;
+  constexpr int nchannels = 3;
+  Roi roi1{0, 3};
+  Roi roi2{{0, 2},
+           {5, 6}};
+  Roi roi3{{0, 0},
+           {0, 0}};
+  Rois rois = {roi1, roi2, roi3};
+
+  std::vector<TensorShape<-1>> ref = {{3, 3, nchannels},
+                                      {4, 5, nchannels},
+                                      {0, 0, nchannels}};
+  auto shs = ShapeFromRoi(rois, nchannels);
+  EXPECT_TRUE(shs == TensorListShape<-1>(ref));
+}
+
+
+TEST(RoiTest, whole_image) {
+  TensorShape<3> sh1 = {3, 3, 5};
+  TensorShape<3> sh2 = {4, 5, 2};
+  TensorShape<3> sh3 = {0, 0, 89};
+
+  Roi ref1 = {0, 3};
+  Roi ref2 = {{0, 0},
+              {5, 4}};
+  Roi ref3 = {0, 0};
+
+  EXPECT_EQ(detail::whole_image<2>(sh1), ref1);
+  EXPECT_EQ(detail::whole_image<2>(sh2), ref2);
+  EXPECT_EQ(detail::whole_image<2>(sh3), ref3);
+}
+
+
+TEST(RoiTest, adjust_empty_roi) {
+  constexpr size_t ndims = 3;
+  std::vector<Roi> rois;
+
+  TensorShape<ndims> ts{5, 6, 7};
+  Roi *roi = nullptr;
+  Roi ref = {{0, 0},
+             {6, 5}};
+  auto res = AdjustRoi(roi, ts);
+  EXPECT_EQ(ref, res);
+}
+
+
+TEST(RoiTest, adjust_roi) {
+  constexpr size_t ndims = 3;
+
+  Roi roi =
+          {{5, 6},
+           {7, 8}};
+  TensorShape<ndims> ts = {12, 13, 14};
+  Roi ref = {{5, 6},
+             {7, 8}};
+  auto res = AdjustRoi(&roi, ts);
+  EXPECT_EQ(ref, res);
+}
+
+
+TEST(RoiTest, adjust_empty_rois) {
+  constexpr size_t ndims = 3;
+  std::vector<Roi> rois;
+
+  TensorListShape<ndims> tls({{2, 3, 4},
+                              {5, 6, 7}});
+  std::vector<Roi> ref = {
+          {{0, 0}, {3, 2}},
+          {{0, 0}, {6, 5}},
+  };
+  auto res = AdjustRoi(rois, tls);
+  ASSERT_EQ(ref.size(), res.size());
+  for (size_t i = 0; i < ref.size(); i++) {
+    EXPECT_EQ(ref[i], res[i]);
+  }
+}
+
+
+TEST(RoiTest, adjust_rois) {
+  constexpr size_t ndims = 3;
+
+  std::vector<Roi> rois = {
+          {{1, 2}, {3, 4}},
+          {{5, 6}, {7, 8}},
+          {0,      20}
+  };
+  std::vector<TensorShape<ndims>> ts = {{9,  10, 11},
+                                        {12, 13, 14},
+                                        {1,  1,  1}};
+  TensorListShape<ndims> tls = ts;
+  std::vector<Roi> ref = {
+          {{1, 2}, {3, 4}},
+          {{5, 6}, {7, 8}},
+          {0,      1}
+  };
+  auto res = AdjustRoi(rois, tls);
+  ASSERT_EQ(ref.size(), res.size());
+  for (size_t i = 0; i < ref.size(); i++) {
+    EXPECT_EQ(ref[i], res[i]);
+  }
+}
+
```

Comment:
add a test case where the Roi is exactly at the edge of the image

Reply:
Doone